### PR TITLE
MultiHdfs final version

### PR DESCRIPTION
**Pull request by pianista215**
## What changes were proposed in this pull request?
Letting to execute through Spark SQL queries that have more than one HDFS kerberized with different configuration

## How was this patch tested?
Tested using the jobs in capsaicinas/demo. TPCDS multitable with different HDFS. And migration Job.

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 1]
```diff
@@ -40,7 +40,7 @@ import org.apache.hadoop.io.{ArrayWritable, BooleanWritable, BytesWritable, Doub
 import org.apache.hadoop.mapred.{FileInputFormat, InputFormat, JobConf, SequenceFileInputFormat, TextInputFormat}
 import org.apache.hadoop.mapreduce.{InputFormat => NewInputFormat, Job => NewHadoopJob}
 import org.apache.hadoop.mapreduce.lib.input.{FileInputFormat => NewFileInputFormat}
-
+import org.apache.hadoop.security.UserGroupInformation
```
Comment (jlopezmalla-stratio) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: If we do not have to add any new Class to this Class we do not have to add those imports, isn't it?

File: core/src/main/scala/org/apache/spark/security/MultiHDFSConfig.scala [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+      // we have a at least a route
+      val pathWithoutHdfs = split.tail.head
+      if (pathWithoutHdfs.startsWith("/")) {
+        // Default HDFS avoid
+        logDebug("Default HDFS")
+        None
+      } else {
+        val hdfsHost = pathWithoutHdfs.split("/").head
+        logDebug(s"Found HDFS HOST: $hdfsHost")
+        Some(hdfsHost)
```
Comment (jlopezmalla-stratio) [inline core/src/main/scala/org/apache/spark/security/MultiHDFSConfig.scala]: IMHO, Option(hdfsHost) would be a more functional solution, do you agree?

File: resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCoarseGrainedSchedulerBackend.scala [hunk 3]
```diff
@@ -730,6 +739,18 @@ private[spark] class MesosCoarseGrainedSchedulerBackend(
...
     }
   }
+
+  /**
+    * Retrieve for all the hosts (multiHdfs) the principal and token associated to this hdfs
+    * @return
+    */
+  override def fetchHadoopDelegationTokens(): Option[Map[String, (String, Array[Byte])]] = {
+    if (UserGroupInformation.isSecurityEnabled) {
+      Some(hadoopDelegationTokenManager.getAllTokens())
```
Comment (jlopezmalla-stratio) [inline resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCoarseGrainedSchedulerBackend.scala]: IMHO, Option(hadoopDelegationTokenManager.getAllTokens()) would be a more functional solution, do you agree?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala [hunk 4]
```diff
@@ -281,15 +282,77 @@ case class FileSourceScanExec(
...
+
+            host ->
+              relation.sparkSession.sparkContext.broadcast(
+                new SerializableConfiguration(finalHadoopConf)
+              )
+        }
+
+      logDebug(s"Creating readFilefunction for files ${relation.inputFiles.mkString("\n")}")
+
+      val host = MultiHDFSConfig.extractHDFSHostFromPath(Some(relation.inputFiles.head))
```
Comment (jlopezmalla-stratio) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: IMHO, Option(relation.inputFiles.head) would be a more functional solution, do you agree?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala [hunk 5]
```diff
@@ -281,15 +282,77 @@ case class FileSourceScanExec(
...
+      }
+
+      // Final function
+      { partitionedFile : PartitionedFile =>
+
+        val filePath = partitionedFile.filePath
+
+        logDebug(s"Read file function for path $filePath")
+
+        val hdfsHost = MultiHDFSConfig.extractHDFSHostFromPath(Some(filePath))
```
Comment (jlopezmalla-stratio) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: IMHO, Option(filePath) would be a more functional solution, do you agree?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala [hunk 6]
```diff
@@ -320,57 +321,68 @@ case class DataSource(
...
-            sparkSession, globbedPaths, options, Some(partitionSchema), fileStatusCache)
-        }
+            val defaultTableSize = sparkSession.sessionState.conf.defaultSizeInBytes
+            new CatalogFileIndex(
+              sparkSession,
+              catalogTable.get,
+              catalogTable.get.stats.map(_.sizeInBytes.toLong).getOrElse(defaultTableSize))
+          } else {
+            new InMemoryFileIndex(
+              sparkSession, globbedPaths, options, Some(partitionSchema), fileStatusCache)
```
Comment (jlopezmalla-stratio) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: IMHO, Option(partitionSchema) would be a more functional solution, do you agree?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala [hunk 7]
```diff
@@ -380,6 +392,26 @@ case class DataSource(
...
   }
 
+  protected def getUgiAndConf(
+                               paths: Iterable[String]
+                             ): (Option[UserGroupInformation], Configuration) = {
+
+    // Support multihdfs So load the config appropiate
+    val hdfsHost = paths.collectFirst {
+      case path if MultiHDFSConfig.extractHDFSHostFromPath(Some(path)).isDefined =>
+        MultiHDFSConfig.extractHDFSHostFromPath(Some(path)).get
```
Comment (jlopezmalla-stratio) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: IMHO, Option(path) would be a more functional solution, do you agree?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala [hunk 8]
```diff
@@ -57,7 +56,9 @@ class InMemoryFileIndex(
   // such streaming metadata dir or files, e.g. when after globbing "basePath/*" where "basePath"
   // is the output of a streaming query.
   override val rootPaths =
-    rootPathsSpecified.filterNot(FileStreamSink.ancestorIsMetadataDirectory(_, hadoopConf))
+    rootPathsSpecified.filterNot( path =>
+      FileStreamSink.ancestorIsMetadataDirectory(path, hadoopConf(Some(path.toUri.toString)))
```
Comment (jlopezmalla-stratio) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala]: IMHO, Option(path.toUri.toString) would be a more functional solution, do you agree?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala [hunk 9]
```diff
@@ -120,9 +121,12 @@ class InMemoryFileIndex(
           pathsToFetch += path
       }
     }
-    val filter = FileInputFormat.getInputPathFilter(new JobConf(hadoopConf, this.getClass))
+    val hdfsHost = MultiHDFSConfig.extractHDFSHostFromPath(Some(paths.head.toUri.toString))
```
Comment (jlopezmalla-stratio) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala]: IMHO, Option(paths.head.toUri.toString) would be a more functional solution, do you agree?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningAwareFileIndex.scala [hunk 10]
```diff
@@ -90,7 +89,8 @@ abstract class PartitioningAwareFileIndex(
       // For each of the root input paths, get the list of files inside them
       rootPaths.flatMap { path =>
         // Make the path qualified (consistent with listLeafFiles and listLeafFilesInParallel).
-        val fs = path.getFileSystem(hadoopConf)
+        val hdfsHost = MultiHDFSConfig.extractHDFSHostFromPath(Some(path.toUri().toString()))
```
Comment (jlopezmalla-stratio) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningAwareFileIndex.scala]: IMHO, Option(path.toUri().toString()) would be a more functional solution, do you agree?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningAwareFileIndex.scala [hunk 11]
```diff
@@ -216,7 +216,9 @@ abstract class PartitioningAwareFileIndex(
   private def basePaths: Set[Path] = {
     parameters.get(BASE_PATH_PARAM).map(new Path(_)) match {
       case Some(userDefinedBasePath) =>
-        val fs = userDefinedBasePath.getFileSystem(hadoopConf)
+        val hdfsHost =
+          MultiHDFSConfig.extractHDFSHostFromPath(Some(userDefinedBasePath.toUri.toString))
```
Comment (jlopezmalla-stratio) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningAwareFileIndex.scala]: IMHO, Option(path.toUri.toString) would be a more functional solution, do you agree?

**Comment by jlopezmalla-stratio** [comment]
Please uncomment the UTTest and check some minnor scala Style changes and ready to merge for me, great job @pianista215 !!

File: core/src/main/scala/org/apache/spark/SparkContext.scala (hunk 1 above)
Comment (pianista215) [reply core/src/main/scala/org/apache/spark/SparkContext.scala]: :+1:

File: core/src/main/scala/org/apache/spark/security/MultiHDFSConfig.scala (hunk 2 above)
Comment (pianista215) [reply core/src/main/scala/org/apache/spark/security/MultiHDFSConfig.scala]: :+1:

File: resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCoarseGrainedSchedulerBackend.scala (hunk 3 above)
Comment (pianista215) [reply resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCoarseGrainedSchedulerBackend.scala]: :+1:

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala (hunk 4 above)
Comment (pianista215) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: :+1:

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala (hunk 5 above)
Comment (pianista215) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: :+1:

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala (hunk 7 above)
Comment (pianista215) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: :+1:

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala (hunk 8 above)
Comment (pianista215) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala]: :+1:

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala (hunk 9 above)
Comment (pianista215) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala]: :+1:

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningAwareFileIndex.scala (hunk 10 above)
Comment (pianista215) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningAwareFileIndex.scala]: :+1:

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningAwareFileIndex.scala (hunk 11 above)
Comment (pianista215) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningAwareFileIndex.scala]: :+1:

**Pull request merged**